trinity/chains/light.py

```diff
@@ -0,0 +1,230 @@
+        self._headerdb = headerdb
+        self._peer_chain = peer_chain
+        self._peer_chain_loop = asyncio.get_event_loop()
+
+    #
+    # Helpers
+    #
+    @classmethod
+    def get_chaindb_class(cls) -> Type[BaseChainDB]:
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    #
+    # Chain API
+    #
+    @classmethod
+    def from_genesis(cls,
+                     base_db: BaseDB,
+                     genesis_params: Dict[str, HeaderParams],
+                     genesis_state: AccountState=None) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    @classmethod
+    def from_genesis_header(cls,
+                            base_db: BaseDB,
+                            genesis_header: BlockHeader) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    def get_chain_at_block_parent(self, block: BaseBlock) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    #
+    # VM API
+    #
+    def get_vm(self, header: BlockHeader=None) -> 'BaseVM':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    @classmethod
+    def get_vm_class_for_block_number(cls, block_number: BlockNumber) -> Type['BaseVM']:
+        """
+        Returns the VM class for the given block number.
+        """
+        validate_block_number(block_number)
+        for start_block, vm_class in reversed(cls.vm_configuration):
+            if block_number >= start_block:
+                return vm_class
+        else:
+            raise VMNotFound("No vm available for block #{0}".format(block_number))
+
+    #
+    # Header API
+    #
+    def create_header_from_parent(self,
+                                  parent_header: BlockHeader,
+                                  **header_params: HeaderParams) -> BlockHeader:
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    def get_block_header_by_hash(self, block_hash: Hash32) -> BlockHeader:
+        return self._headerdb.get_block_header_by_hash(block_hash)
+
+    def get_canonical_head(self):
+        return self._headerdb.get_canonical_head()
+
+    #
+    # Block API
+    #
+    def get_ancestors(self, limit: int) -> Iterator[BaseBlock]:
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    def get_block(self) -> BaseBlock:
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    def get_block_by_hash(self, block_hash: Hash32) -> BaseBlock:
+        try:
+            header = self._headerdb.get_block_header_by_hash(block_hash)
+        except HeaderNotFound:
+            header = self._run_async(
+                self._peer_chain.get_block_header_by_hash(block_hash)
```

Comment:
This looks fishy to me.  Should we allow header retrieval for headers that aren't in the headerdb, knowing that the light client does have a full db of the header chain?

Reply:
The only win I could think of is that it allows us to access headers that we haven't synced to yet, but I don't think it's a critical feature (at least right now). I'll drop it.